### ISSUE TITLE
Variable length FORMAT fields (Number=.) not supported

### ISSUE THREAD

**Issue by dancooke** [samtools/htslib#664]
As far as I understand, the VCF specification ([v4.3](https://github.com/samtools/hts-specs/blob/master/VCFv4.3.pdf)) permits FORMAT fields with variable length. In particular, section 1.4.2 (Information field format) says about the "Number" entry:

    . (dot): The number of possible values varies, is unknown or unbounded.

Section 1.4.4 (Individual format field format) goes on to say:

    Genotype fields specified in the FORMAT field are described as follows:
    ##FORMAT=<ID=ID,Number=number,Type=type,Description="description">
    Possible Types for FORMAT fields are: Integer, Float, Character, and String (this field is otherwise defined precisely as the INFO field).

Section 1.6.2 (Genotype fields) has more to say about FORMAT fields, but the only relevant statement seems to be:

    If genotype information is present, then the same types of data must be present for all samples.
    First a FORMAT field is given specifying the data types and order (colon-separated FORMAT ids matching the regular expression ^[A-Za-z ][0-9A-Za-z .]*$, duplicate fields are not allowed).
    This is followed by one data block per sample, with the colon-separated data corresponding to the types specified in the format.

So taken together, I believe a field such as

     ##FORMAT=<ID=FTS,Number=.,Type=String,Description="Filter status">

are valid. On another note, I think the reserved `FT` FORMAT field should have `Number=.`.

It appears htslib, at least for parsing, is not able to handle FORMAT fields with variable length. In particular, the method `bcf_update_format` in `vcf.h` contains the following:

```c
/*
     * @values:    pointer to the array of values, the same number of elements
     *             is expected for each sample. Missing values must be padded
     *             with bcf_*_missing or bcf_*_vector_end values.
     *  @n:        number of values in the array. If n==0, existing tag is removed.
*/
int bcf_update_format(const bcf_hdr_t *hdr, bcf1_t *line, const char *key, const void *values, int n, int type)
{
....
    line->n_sample = bcf_hdr_nsamples(hdr);
    int nps = n / line->n_sample;  // number of values per sample
    assert( nps && nps*line->n_sample==n );     // must be divisible by n_sample
....
}
```

I haven't looked any further than this, so I don't know if htslib uses a data structure that is capable of storing variable length fields or not.

**Comment by pd3** [comment]
FORMAT string of variable length are passed to the function as regular comma-separated strings.

**Comment by dancooke** [comment]
What about non-string types? This seems like half-baked support to me. I don't see this documented anywhere.

**Comment by pd3** [comment]
Well, some prefer medium, others medium rare. For speed, the returned data reflects the structure of BCF. Each FORMAT annotation is returned as an array of fixed-length per-sample blocks. In case of variable number of fields, different samples can be of different lengths, shorter blocks are padded with end-of-vector bytes.

**Comment by dancooke** [comment]
Padding works for strings, but what about other types?

**Comment by dancooke** [comment]
Well, padding doesn’t entirely work for strings either - what about empty strings...

**Comment by lh3** [comment]
Other types are the same: padded with end-of-vector values.

**Comment by dancooke** [comment]
That's not safe, what if I want to assign a value of zero for the last 1+ value(s)?

**Comment by lh3** [comment]
I don't quite understand your question. Will try with my own interpretation – BCF keeps the max number of each FORMAT field, so htslib will not decode more than that number.

Also, do you know the memory layout of a VCF record? If not, you need to understand that first. I am not sure where this is documented nowadays, though.

**Comment by lh3** [comment]
Oh, also, end-of-vector (EOV) value for float arrays is 0x7F800002 (4 bytes). EOV of 8-bit integer array is -127. See also `_vector_end` in vcf.h.

**Comment by daviesrob** [comment]
`bcf_update_format()` takes a block of values which contains all the data values for the first sample then all the data for the second sample (if present) and so on.  In this block the number of values for each sample must be the same, so where the number of values per sample varies, the shorter ones must be padded out to the length of the longest with the appropriate vector_end value defined in `htslib/vcf.h`.
`bcf_update_format()` expects 32-bit integers, so this would be `bcf_int32_vector_end` for integers and `bcf_float_vector_end` for floating-point values.

If you want to encode missing values in the middle of the list then you should use `bcf_int32_missing` for integers or `bcf_float_missing` for floating-point.  Missing values and vector end both encode the absence of data, the difference is that a loop through the data should keep going if it finds a missing value but can terminate early if it finds vector end.

Strings are slightly different.  You pass in one string per sample and separate values within that string using commas (as already mentioned).  All of the strings have to be the same length, so again the shorter ones have to be padded out to the length of the longest.  In this case you use set the unused bytes to zero.  In the case of missing data in theory `bcf_str_missing` (actually a tab character) should be used used for the first character, although `vcf_parse()` seems to just copy whatever was in the vcf file (presumably '.').  Either way, the rest of the space is padded with zero bytes.

The parameter `n` passed to `bcf_update_format()` is the total number of values, including missing and end of vector padding.  For strings this is the total length of all the strings including padding.

So in summary, it works by padding the data for each sample out to the same length.  And while this length may vary, you can always work out what it is for a given BCF record by counting the number of values you actually have.

**Comment by dancooke** [comment]
Ah I see, I hadn't seen in the VCF spec that certain FORMAT values are reserved. Thanks for the explanation. I guess it is what it is, but I can see that being the source of quite a lot of confusion/bugs; I can't imagine many people are actually checking the values that go into these fields. I also wonder the real term gains of using padding versus explicitly storing lengths in the variable case...
